### DISCUSSION TITLE
Having issues trying to get my eventhandler to run before or after a built in command

### DISCUSSION THREAD

**Discussion by jimlocigno** [VsixCommunity/Community.VisualStudio.Toolkit#478] [Q&A]
I have tried many things, searched google, looked at Mads extension library - found something kind of close there - but this still isn't working for me at least.

I eventually have certain commands in mind that I want to execute my own code either before or after but no matter what approach I take my custom code does not seem to execute? I'm an experienced developer but very new to VSIX.

Everything  I find, points me to a different way of executing custom code before or after a given command but nothing seems to work?  

I set breakpoints in my eventhandlers below and they never get hit.

_Anyhelp would be greatly appreciated!_

Here is my latest try:
`
namespace VSIXProject13
{
    [PackageRegistration(UseManagedResourcesOnly = true, AllowsBackgroundLoading = true)]
    [InstalledProductRegistration(Vsix.Name, Vsix.Description, Vsix.Version)]
    [ProvideMenuResource("Menus.ctmenu", 1)]
    [Guid(PackageGuids.VSIXProject13String)]
    public sealed class VSIXProject13Package : ToolkitPackage
    {

        public const string PackageGuidString = "bcf67d7a-8e10-4fe5-af47-2fa32560656a";
        private DTE2 _dte;

        protected override async Task InitializeAsync(CancellationToken cancellationToken, IProgress<ServiceProgressData> progress)
        {
            await this.RegisterCommandsAsync();

            await JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);

            _dte = (DTE2)await GetServiceAsync(typeof(DTE));
            Assumes.Present(_dte);
       
            // Hook up event handlers for before and after command execution
            _dte.Events.DTEEvents.OnBeginShutdown += OnBeginShutdown;
            _dte.Events.CommandEvents.BeforeExecute += OnBeforeCommandExecute;
            _dte.Events.CommandEvents.AfterExecute += OnAfterCommandExecute; _dte = (DTE2)GetService(typeof(DTE));

        }

    private void OnBeforeCommandExecute(string Guid, int ID, object CustomIn, object CustomOut, ref bool CancelDefault)
        {
            var sw = new StreamWriter(@"c:\test\vsix.log");
            sw.AutoFlush = true;
            sw.WriteLine("here-1\n");
            sw.Close();
            // Add custom code to run before the built-in command execution
            // Example: Console.WriteLine("Before executing command: " + _dte.Commands.Item(Guid, ID).Name);
        }

        /// <summary>
        /// Event handler for after command execution.
        /// </summary>
        private void OnAfterCommandExecute(string Guid, int ID, object CustomIn, object CustomOut)
        {
            // Add custom code to run after the built-in command execution
            // Example: Console.WriteLine("After executing command: " + _dte.Commands.Item(Guid, ID).Name);
            Debug.Print("We are in the OnAfterCommandExecute event");
        }

        /// <summary>
        /// Event handler for when the IDE begins to shut down.
        /// </summary>
        private void OnBeginShutdown()
        {
            // Clean up resources and unregister event handlers
            _dte.Events.DTEEvents.OnBeginShutdown -= OnBeginShutdown;
            _dte.Events.CommandEvents.BeforeExecute -= OnBeforeCommandExecute;
            _dte.Events.CommandEvents.AfterExecute -= OnAfterCommandExecute;
        }
    }
}`

**Comment by madskristensen** [comment]
You need to add the `[ProvideAutoload]` attribute to the package class. Otherwise the `InitializeAsync` method will never be executed by VS
